### PR TITLE
meta: use direnv

### PR DESCRIPTION
**Pull request by xwvvvvwx**
- Support `lorri` in `shell.nix`
- Add a `.envrc` and use `lorri` if available

### CODE REVIEW

File: shell.nix [hunk 1]
```diff
@@ -39,8 +39,8 @@ pkgs.stdenv.mkDerivation {
     z3
   ];
   shellHook = ''
-    export PATH=$PWD/node_modules/.bin/:$PWD/bin:$PATH
-    export KLAB_EVMS_PATH="''${KLAB_EVMS_PATH:-''${PWD}/evm-semantics}"
+    export PATH=${toString ./node_modules/.bin}:${toString ./bin}:$PATH
```
Comment (asymmetric) [inline shell.nix]: Haha I had started working on this PR and then got distracted trying to use a nix function to join strings by `:`, but I don't think that's really necessary.

File: shell.nix [hunk 2]
```diff
@@ -39,8 +39,8 @@ pkgs.stdenv.mkDerivation {
...
+    export KLAB_EVMS_PATH="''${KLAB_EVMS_PATH:-${toString ./evm-semantics}}"
```
Comment (asymmetric) [inline shell.nix]: Do we really need to do this default-argument thing?

If this file is sourced directly, we can set the var to its desired value. If we're depending upon klab in a parent project, the variable will be assigned to in the parent's `shell.nix` (and this one will not be sourced).

Are there cases where we source this file while having already set this variable, and are they legitimate?

I'm also happy to do it in another PR if you prefer.

**Comment by livnev** [comment]
Can someone ELI5 the difference between `lorri` and just `use nix` in `.envrc` (the latter is what I use currently).

File: shell.nix (hunk 2 above)
Comment (xwvvvvwx) [reply shell.nix]: > Are there cases where we source this file while having already set this variable, and are they legitimate?

Good question. I guess the use case would be to quickly try out different `evms` versions with the same `klab`, but that can also be achieved by checking out branches in the submodule.

I think if we remove the default arg stuff here we should consider removing the variable entirely for the same reason?

**Comment by xwvvvvwx** [comment]
[`lorri`](https://github.com/target/lorri) is a nix-shell replacement that:

- caches environments (no more multi second wait to evaluate the `nix-shell`)
- automatically creates garbage collection roots for the shells that it manages (no more waiting to redownload the world after a gc run)

**Comment by livnev** [comment]
> [`lorri`](https://github.com/target/lorri) is a nix-shell replacement that:
> 
> * caches environments (no more multi second wait to evaluate the `nix-shell`)
> * automatically creates garbage collection roots for the shells that it manages (no more waiting to redownload the world after a gc run)

I see. I have [this copypasta](https://lev.liv.nev.org.uk/pub/038ac788e2019633674c182575b99c10ed57085d54255a3fcc26b4d193669f2e) in my `.direnvrc`, which I think attempts to do at least the first thing you mentioned? Is this deprecated in favour of `lorri` now?

**Comment by xwvvvvwx** [comment]
`lorri` is probably the recommended solution nowadays. It's written in :sparkles: rust :sparkles: after all...

File: shell.nix (hunk 2 above)
Comment (asymmetric) [reply shell.nix]: > Good question. I guess the use case would be to quickly try out different evms versions with the same klab, but that can also be achieved by checking out branches in the submodule.

Yeah, so far I've used git (branches and worktrees) to test out different versions of evm/klab/whatever. That's a bit more convoluted but it's also more general and doesn't require special support.

> I think if we remove the default arg stuff here we should consider hardcoding the value within klab itself for the same reason?

Yeah good point. We could remove both, unless keeping it would make more sense for non-nix users, of which I'm not sure.

Let's move this to a new PR, since touching klab proper would be out of scope.

Comment (xwvvvvwx) [reply shell.nix]: > Let's move this to a new PR, since touching klab proper would be out of scope.

:+1:
